image-hd.c

```diff
@@ -143,13 +153,36 @@ static int hdimage_insert_mbr(struct image *image, struct list_head *partitions)
 			break;
 		i++;
 	}
+
+	if (hybrid) {
+		struct mbr_partition_entry *entry;
+
+		entry = (struct mbr_partition_entry *)(part_table + i *
+			sizeof(struct mbr_partition_entry));
+
+		entry->boot = 0x00;
+
+		entry->partition_type = 0xee;
+		entry->relative_sectors = 1;
+		entry->total_sectors = (512+sizeof(struct gpt_header) +
+			(sizeof(struct gpt_partition_entry) * GPT_ENTRIES))/512;
```

Comment:
This should cover the all gpt entries, right? They may not be located directly after the header if `gpt-location` is set.
I've just created #99. This fixes the same problem elsewhere.
Also, this is the size, not the position, so I think `hd->gpt_location + (GPT_SECTORS - 1) * 512 - 512` is probably correct.`

Reply:
> This should cover the all gpt entries, right?

No, this only covers the primary GPT table region(the area between the MBR and the first partition) and not any actual real partition entries. Since we need the first partition(fat32 boot partition in the case of the pi) to be usable when read by software that only understands MBR partition tables we can't overlap the GPT partitions like a protective MBR. Protective MBR is designed to show the entire disk as already formatted/used when read in MBR mode which is not what we want here as we need to actually be able to mount partitions(such as the boot partition) in MBR mode.

> Also, this is the size, not the position,

I know, size should be correct as entry start position(start of primary GPT partition table) and primary GPT partition table size is always the same from my understanding.

Here's what it looks like for GPT:
```
$ fdisk -t gpt -l rpi3_20200326_055346.img
Disk rpi3_20200326_055346.img: 10.3 GiB, 10771006976 bytes, 21037123 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 89EFFCC0-9035-49AB-B653-F752D71D8709

Device                    Start      End  Sectors Size Type
rpi3_20200326_055346.img1    34    65569    65536  32M EFI System
rpi3_20200326_055346.img2 65570 21037089 20971520  10G Linux root (ARM-64)
```
and in MBR mode for the same disk:
```
$ fdisk -t dos -l rpi3_20200326_055346.img
Disk rpi3_20200326_055346.img: 10.3 GiB, 10771006976 bytes, 21037123 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x00000000

Device                    Boot Start   End Sectors  Size Id Type
rpi3_20200326_055346.img1 *       34 65569   65536   32M  c W95 FAT32 (LBA)
rpi3_20200326_055346.img3          1    33      33 16.5K ee GPT

Partition table entries are not in disk order.
```

> I've just created #99. This fixes the same problem elsewhere.

I'm not entirely sure I'm fully understanding the implications here but I know we need the protective MBR part of the hybrid MBR to start covering at sector 1 and end before our first real partition as we are only protecting the GPT header/table not the actual partitions.

Reply:
To clarify a hybrid MBR partition needs the 0xee partition because software that understands GPT partitions knows to operate in GPT mode when it sees the 0xee MBR partition entry. So in the case of the raspberry pi the bootloader bootstraps and launches the kernel using the MBR partition table to access the fat32 boot partition. When the kernel launches and reads the disk it sees the 0xee partition tag and switches to GPT mode automatically and mounts the rootfs using the GPT partition table entry and not the MBR entry. The MBR partition table entries are intentionally not in disk order so that the firmware just needs to look for the first entry to bootstrap/launch the kernel.

Reply:
> I'm not entirely sure I'm fully understanding the implications here but I know we need the protective MBR part of the hybrid MBR to start covering at sector 1 and end before our first real partition as we are only protecting the GPT header/table not the actual partitions.

I understand that. And that is what the code I proposed is doing.

The situation we have is this. The primary GPT table consists of the GPT header (always the second sector) and the GPT table. Usually, this list start in the third sector directly after the GPT header.
However, this is not necessary. Sometimes, some space is needed for a bootloader between the GPT header and the table. In this case, `gpt_location` is set to something other than `2*512`.
This must be taken into account when calculating the size of the special entry in the protective MBR:
`GPT_SECTORS` is the number of sectors for header + table, so the table ends at `gpt_location + (GPT_SECTORS - 1)*512`. The extra `-512` is to get the size because we start at the second sector.

Reply:
Hmm, something with the math must be off as the 0xee protective partition was going past the start of the first real partition when I tried building with your suggested change.